src/Chapter3.hs

```diff
@@ -479,6 +551,52 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
+data Castle = NoCastle | CityCastle String deriving (Show)
+
+data Walls = NoWalls | CityWalls deriving (Show)
+
+data Hall = Church | Library deriving (Show)
+
+data House = BuildHouse 
+    {
+      housePoeple :: Int
+    } deriving (Show)
+
+data City = BuildCity
+    { cityCastle :: Castle
+    , cityHall :: Hall
```

Comment:
This is the correct data type, that can represent what we have in requirements 👍🏼 
However, this also allows having some states that should not ideally exist: like having walls without any castle at all. With this representation, this is possible. However, we actually can restrict this on type-level by providing more representative types 🙂 This is actually the greatest power of Haskell and type system 😍 

This pattern is called "Make illegal states unrepresentable", and you can read more about it here:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable